src/index.d.ts

```diff
@@ -0,0 +1,55 @@
+import React from 'react';
+
+type BoxOptions = 'border-box' | 'content-box' | 'device-pixel-content-box';
+
+interface ResizeObserverBoxSize {
+  readonly inlineSize: number;
+  readonly blockSize: number;
+}
+
+interface ResizeObserverEntry {
+  readonly target: Element;
+  readonly contentRect: DOMRectReadOnly;
+  readonly borderBoxSize: Array<ResizeObserverBoxSize> | ResizeObserverBoxSize;
+  readonly contentBoxSize: Arrray<ResizeObserverBoxSize> | ResizeObserverBoxSize;
+  readonly devicePixelContentBoxSize: Arrray<ResizeObserverBoxSize> | ResizeObserverBoxSize;
+}
+
+interface Breakpoints<T> {
+  [key: number]: T;
+}
+
+interface BreakpointsOptions<W, H> {
+  box?: BoxOptions;
+  widths?: Breakpoints<W>;
+  heights?: Breakpoints<H>;
+  fragment?: number;
+}
+
+interface ObservedElementProps<T> {
+  ref: React.RefObject<T>;
+}
+
+interface ObserveRenderArgs<E, W, H> {
+  observedElementProps: ObservedElementProps<E>;
+  widthMatch: W;
+  heightMatch: H;
+}
+
+interface ObserveProps<E, W, H> {
+  box?: BoxOptions;
+  breakpoints?: BreakpointsOptions<W, H>;
+  render: ({
+    observedElementProps,
+    widthMatch,
+    heightMatch
+  }: ObserveRenderArgs<E, W, H>) => JSX.Element;
+}
+
+export const Observe: <E, W, H>(props: ObserveProps<E, W, H>) => JSX.Element;
+
+export const useBreakpoints: <W extends any, H extends any>(options: BreakpointsOptions<W, H>) => [W, H];
```

Comment:
This hook accepts an options object with this shape:
```javascript
options = {
  box: 'border-box' | 'content-box' | 'device-pixel-content-box',
  widths: {
    0: any,
    769: any
  },
  heights: {
    0: any,
    1080: any
  },
  fragment: 0
};
```
All keys in `options` are optional, but a user should provide at least **either** `widths` or `heights`. Not specifying both will cause `useBreakpoints()` to always return `[undefined, undefined]`. I don't know how I can communicate this soft constraint in TypeScript.

Inside the `widths` and `heights` objects, the keys are always numeric **and should always increase**. I don't know how to communicate this constraint in TypeScript. Specifying a lower key after specifying a higher key will cause the former to be missed by the hook's logic. I'm not re-ordering the `widths` and `heights` objects internally, nor do I think I should.

`box` and `fragment` are optional, and only work when your `ResizeObserver` supports the [draft spec](https://drafts.csswg.org/resize-observer/#api) box sizes and fragments.

`box` can be one of `undefined`, `'border-box'`, `'content-box'`, `'device-pixel-content-box'`.

`fragment` can be `undefined` or a 0-based positive integer.

This hook will return an array with two values.
```javascript
const [widthMatch, heightMatch] = useBreakpoints(...)
```
Those values are the value belonging to the matching `widths` key and `heights` key respectively. The type of this value can be anything. **I'm even allowing the value to be of a _different_ type for each key in `widths` and `heights`.** This was probably the hardest to write a type definition for, and your feedback would be greatly appreciated.